lib/saluki-io/src/deser/codec/dogstatsd.rs

```diff
@@ -551,20 +581,156 @@ impl<'a> Iterator for ValueIter<'a> {
+    }
+}
+
+#[derive(Debug, Clone)]
+struct TagFilterer<I, TMI> {
+    iter: I,
+    interceptor: TMI,
+}
+
+impl<I, TMI> TagFilterer<I, TMI> {
+    fn new(iter: I, interceptor: TMI) -> Self {
+        Self { iter, interceptor }
+    }
+}
+
+impl<'a, I, TMI> IntoIterator for TagFilterer<I, TMI>
+where
+    I: IntoIterator<Item = &'a str> + Clone,
+    TMI: TagMetadataInterceptor,
+{
+    type Item = I::Item;
+    type IntoIter = TagFiltererIter<I::IntoIter, TMI>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        TagFiltererIter {
+            iter: self.iter.into_iter(),
+            interceptor: self.interceptor,
+        }
+    }
+}
+
+struct TagFiltererIter<I, TMI> {
+    iter: I,
+    interceptor: TMI,
+}
+
+impl<'a, I, TMI> Iterator for TagFiltererIter<I, TMI>
+where
+    I: Iterator<Item = &'a str>,
+    TMI: TagMetadataInterceptor,
+{
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match self.iter.next() {
+                Some(tag) => match self.interceptor.evaluate(tag) {
+                    InterceptAction::Pass => return Some(tag),
+                    InterceptAction::Intercept | InterceptAction::Drop => continue,
+                },
+                None => return None,
+            }
+        }
+    }
+}
+
+fn update_metadata_from_tags<'a, I, TMI>(tags_iter: I, interceptor: &TMI, metadata: &mut MetricMetadata)
+where
+    I: Iterator<Item = &'a str>,
+    TMI: TagMetadataInterceptor,
+{
+    for tag in tags_iter {
+        if let InterceptAction::Intercept = interceptor.evaluate(tag) {
+            interceptor.intercept(tag, metadata)
+        }
+    }
+}
+
 #[cfg(test)]
 mod tests {
+    use std::sync::Arc;
+
     use nom::IResult;
     use proptest::{collection::vec as arb_vec, prelude::*};
     use saluki_context::{ContextRef, ContextResolver};
+    use saluki_core::{pooling::helpers::get_pooled_object_via_default, topology::interconnect::EventBuffer};
     use saluki_event::{metric::*, Event};
 
-    use super::{parse_dogstatsd, DogstatsdCodecConfiguration};
+    use crate::deser::{codec::DogstatsdCodec, Decoder};
+
+    use super::{parse_dogstatsd, DogstatsdCodecConfiguration, InterceptAction, TagMetadataInterceptor};
 
     enum OneOrMany<T> {
         Single(T),
         Multiple(Vec<T>),
     }
 
+    #[derive(Debug)]
+    struct StaticInterceptor;
+
+    impl TagMetadataInterceptor for StaticInterceptor {
+        fn evaluate(&self, tag: &str) -> InterceptAction {
+            if tag.starts_with("host") {
+                InterceptAction::Intercept
+            } else if tag.starts_with("deprecated") {
+                InterceptAction::Drop
+            } else {
+                InterceptAction::Pass
+            }
+        }
+
+        fn intercept(&self, tag: &str, metadata: &mut MetricMetadata) {
+            if let Some((key, value)) = tag.split_once(':') {
+                if key == "host" {
+                    metadata.set_hostname(Arc::from(value));
```

Comment:
unrelated question: why is `hostname` the only field the needs to be wrapped in an `Arc`for `MetricMetadata`?

Reply:
It's the only string field, and to make `MetricMetadata` cloneable, as least cheaply... we wrap the hostname in `Arc<T>`.

Realistically, it should actually be `MetaString` so that we can intern it or inline it where possible, skipping a heap allocation entirely.